### PR TITLE
Fix `--inputs` 2 — `invoke` command

### LINKED ISSUE #1220: Fix handling calldata provided to `protostar invoke`
> I have to provide 2 inputs, a uint256 and a struct array (struct is (felt, felt)), so something like that: —inputs uint256 array_len array_values. With an array of length 1, it is supposed to look like that: —inputs int int int int. The problem is that it keeps returning me that error `Provided 4 positional arguments, 2 possible..`

### PR DESCRIPTION
**Pull request by kasperski95**
Fixes #1220

### CODE REVIEW

File: protostar/starknet_gateway/data_transformer.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            return None
+        transform = to_python_transformer(
+            contract_abi=abi,
+            fn_name=str(selector),
+            mode="outputs",
+        )
+        return transform(data)
+
+
+class AbiNotFoundException(ProtostarException):
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/data_transformer.py:68]: Do we need this exception type? We don't catch it anywhere

File: protostar/starknet/types.py [hunk 2]
```diff
@@ -1,2 +1,9 @@
+from typing import Union
+
 SelectorType = int
 ClassHashType = int
+
+CairoDataRepresentation = list[int]
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet/types.py:6]: Those types belong to `protostar/starknet_gateway/data_transformer.py`, they are not related to starknet os. `Wei` can stay here

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]

File: protostar/commands/invoke/messages.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+from dataclasses import dataclass
+from typing import Optional
+
+from protostar.io.log_color_provider import LogColorProvider
+from protostar.io.output import StructuredMessage
+from protostar.starknet_gateway.invoke import InvokeOutput
+
+
+class SendingInvokeTransactionMessage(StructuredMessage):
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/invoke/messages.py:9]: This name conflicts with the message type (line 15).

File: protostar/starknet_gateway/data_transformer.py (hunk 1 above)
Comment (kasperski95) [reply protostar/starknet_gateway/data_transformer.py:68]: It can be replaced with `ProtostarException`.

File: protostar/starknet/types.py (hunk 2 above)
Comment (kasperski95) [reply protostar/starknet/types.py:6]: They are related to StarkNet. For example, the testing cheatcode `deploy_contract` uses data transformer, but it doesn't use StarkNet client.

File: protostar/starknet_gateway/data_transformer.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        if abi is None:
+            raise AbiNotFoundException(
+                message=(
+                    f"Couldn't resolve ABI for address: {address}.\n"
+                    "Provide ABI to use data transformer."
+                )
+            )
+        return abi
+
+    async def try_transforming_entrypoint_output_to_human(
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/data_transformer.py:50]: Why should it throw an exception?

File: protostar/starknet_gateway/invoke/invoke_use_case.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    async def execute(self, input_data: InvokeInput) -> InvokeOutput:
+        cairo_calldata = (
+            await self._data_transformer.transform_entrypoint_input_to_cairo(
+                address=input_data.address,
+                selector=input_data.selector,
+                calldata=input_data.calldata,
+                abi=input_data.abi,
+            )
+        )
+        unsigned_invoke_tx = UnsignedInvokeTransaction(
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/invoke/invoke_use_case.py:33]: I was going to remove it to be consistent with changes in the previous PR. That's why this PR is in draft.

Dataclasses reduce the boilerplate when used with Protocols, and they allow creating shared dataclasses without changing function signatures.

File: protostar/starknet_gateway/core/structs.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+from dataclasses import dataclass
+from typing import Optional
+
+from protostar.starknet import Address, CairoDataRepresentation
+
+from .types import Wei
+
+
+@dataclass
+class TransactionSentResponse:
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/core/structs.py:10]: Ok

File: protostar/commands/invoke/messages.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/invoke/messages.py:9]: True, hmm maybe `Pending`? idk Sending sounds weird, leaving it up to you as it is [[Nitpick]]

File: protostar/starknet_gateway/data_transformer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/starknet_gateway/data_transformer.py:50]: `try_...` is cryptic, you look at the function and you have to go inside to check what does it even mean.

if you do 
```
try:
 do_something()
 except abi not found
  pass
```
you instantly understand everything. Does it make sense?

File: pyproject.toml [hunk 7]
```diff
@@ -63,7 +63,7 @@ ci = [
 custom_checks = "sh ./scripts/custom_checks.sh"
 deploy = "python ./scripts/deploy.py"
 deploy_prerelease = "python ./scripts/deploy_prerelease.py"
-devnet = "starknet-devnet"
+devnet = "starknet-devnet --seed 0"
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply pyproject.toml:66]: Creating a separate PR for this change is an overkill.

File: protostar/starknet_gateway/invoke/invoke_use_case.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/starknet_gateway/invoke/invoke_use_case.py:33]: >That's why this PR is in draft.
It is open :p
Screenshot 2022-12-21 at 12 06 47
IDK maybe by mistake

File: protostar/starknet_gateway/data_transformer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/data_transformer.py:50]: I'll rename it to the `transform_entrypoint_output_to_human_if_abi_found()`.

File: protostar/starknet_gateway/invoke/invoke_use_case.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/invoke/invoke_use_case.py:33]: Sorry, my bad.

File: protostar/starknet/types.py (hunk 2 above)
Comment (kasperski95) [reply protostar/starknet/types.py:6]: The data transformer core is in `protostar/starknet/data_transformer.py`.

File: protostar/starknet_gateway/data_transformer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/starknet_gateway/data_transformer.py:50]: Sound good

File: protostar/commands/invoke/messages.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/invoke/messages.py:9]: Left it as it is to preserve consistency with the message type. I'd change it to pending if that word would be used in the message to the user, but it sounds weird there.

File: protostar/starknet_gateway/data_transformer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/data_transformer.py]: Renamed

File: protostar/starknet_gateway/invoke/invoke_use_case.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/invoke/invoke_use_case.py]: Removed

File: protostar/starknet_gateway/account_manager.py [hunk 8]
```diff
@@ -69,13 +80,45 @@ async def sign_multicall_transaction(
...
         if actual_account_version != self._account_client.supported_tx_version:
             raise UnsupportedAccountVersionException(actual_account_version)
 
+    async def execute(
+        self, payload: PayloadToAccountExecuteInvokeTx
+    ) -> TransactionSentResponse:
+        return await self._client.send_payload_to_account_execute(payload)
+
 
 class UnsupportedAccountVersionException(ProtostarException):
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/account_manager.py:126]: Removed

File: protostar/starknet_gateway/core/structs.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/core/structs.py]: Replaced

**Review by MaksymilianDemitraszek — APPROVED** [review]
Please rename DataTransformer -> DataTransformationPolicy and reuse CairoData, PythonData

File: protostar/starknet/types.py (hunk 2 above)
Comment (kasperski95) [reply protostar/starknet/types.py]: Internal discussion. Renamed `DataTransformer` to `DataTransformerPolicy`.

**Pull request merged**